RFC 117: Use a container runtime (#117)
alphagov/govuk-rfcs: rfc-117-containerised-runtime.md

> # RFC 000: Use a containerised runtime

Reviewer: ```suggestion
# RFC 117: Use a containerised runtime
```

## Summary

> Use a containerised runtime for hosting GOV.UK applications. GOV.UK applications currently run on virtual machines running Ubuntu Trusty Tahr. There is currently an extended support contract for Ubuntu Trusty which will end in April 2022. Rather than upgrading Ubuntu on the virtual machines GOV.UK should change the application runtime to a containerised platform.

Reviewer 2: @theseanything pointed out that there's too much emphasis in this summary on the Trusty upgrade - I agree. This RFC is wider than just "how do we get of EOL ubuntu". Suggest we change it to something like:

```suggestion
Use a containerised runtime for hosting GOV.UK applications. This will provide a number of benefits, including:
* Removing our dependency on a version of Ubuntu (Trusty) which is in extended support (which ends in April 2022)
* Removing legacy infrastructure code which is difficult to maintain and extend
* Providing better value for money by using cloud resources more effectively
* Improving development velocity by reducing the difference between production and development environments
* Making it easier to adopt higher velocity deployment approaches (such as continuous deployment)
```

## Problem

GOV.UK currently hosts the majority of its applications and services using virtual machines running Ubuntu 14.04 LTS (Trusty Tahr).

Trusty is out of date, it was released in April 2014 and standard support ended in April 2019. At this time GOV.UK signed up for [Extended Security Maintenance](https://ubuntu.com/esm) (ESM) which provides continued security fixes for high and critical common vulnerabilities and exposures (CVEs) for the [most commonly used packages](https://wiki.ubuntu.com/SecurityTeam/ESM/14.04#A14.04_Infrastructure_ESM_Packages). ESM will end in April 2022.

> Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).

Reviewer 3: We're running MongoDB 3.2 on Ubuntu, as well as 2.4, so I'm not sure the reason why we're stuck using 2.4 for many apps is because of using an older version of Ubuntu (although I realise this is a factor in general).

Reviewer 2: Mongo 4.2 claims to [drop support for Trusty](https://docs.mongodb.com/manual/tutorial/install-mongodb-on-ubuntu/#platform-support). Not that we necessarily _want_ to upgrade to Mongo 4.2 - we might choose to move to managed Mongo like DocumentDB, which would avoid this. Still though, being stuck on Trusty is narrowing our options.

> Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:

Reviewer: If I remember correctly one of the blockers (if not the main one) of upgrading to Xenial was that Icinga wouldn't run on Xenial. Not sure if that is still an issue.

Reviewer 2: It looks like Icinga has [Xenial binaries](https://packages.icinga.com/ubuntu/dists/icinga-xenial/) now, but I don't know if there was another reason why it wouldn't run - perhaps we'd need to do a big Icinga upgrade to use those binaries?

I don't think we need to call out every blocker in the RFC though - it will end up being huge!

- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.

> - GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.

Reviewer: Presumably moving to a containerised system requires an even greater rewrite/rethink of these than the switch to systemd? so this may be a problem in all scenarios.

Reviewer 4: Do you have an example @kevindew? I assume whatever container management system we use would essentially replaces upstart (which is probably out of scope for now)

Reviewer 3: I think Kevin's point is that switching from Upstart to a "container runtime" for process management is a bigger switch than switching from Upstart to systemd, as systemd is more similar to upstart than container runtimes, as that's a different but related concept.

Reviewer: Yup, @cbaines summarised it well 👍

Reviewer 2: Yeah, the container runtime does replace the init system, but it's true that we'll still need to do the work to make sure that the applications behave reliably in the new world. I guess this work will be quite similar in nature to the switch to systemd, but at least we won't have to do it with puppet 😅.

I still think it's worth calling this out as a thing that makes upgrading past Trusty difficult, even though the solution in the RFC doesn't make the work go away entirely.

Reviewer 5: > GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.

I agree with Kevin and Chris's points. It might be useful to have an example in the RFC to demonstrate how significant rewriting these scripts would be.

Reviewer 2: I'm not sure I can actually support this statement. In theory puppet supports both upstart and systemd, and it should magically handle it all for us. There are a [few references to upstart specifically in govuk-puppet](https://github.com/alphagov/govuk-puppet/search?q=upstart&unscoped_q=upstart), but not so many that it looks like it would be a massive job.

In practice I think it would be a bit harder than that, but without trying it I can't know for sure. I absolutely do not want to try it myself 🙅‍♂.

Maybe let's just remove this point since it's controversial?

Reviewer 2: I've watered the point down, rather than remove it, in af3889bda2efd9c68a8a6c6f1ab05c67e3c5c293.

I don't know enough about the systemd upgrade to come up with concrete examples of why it would be hard, but my spidey senses are definitely tingling with danger.

- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.

## Proposal

> GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.

Reviewer 2: ```suggestion
GOV.UK will use a managed, container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
```

Suggesting this amendment to address @cbaines's comment. Paraphrasing, to "adopt" a platform has connotations of "taking responsibility for looking after the platform", whereas I think our intent here is to use a managed platform, precisely so we can avoid having to manage lower-level infrastructure concerns.

> GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.

Reviewer 4: ```suggestion
GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
```
Not sure its clear or accurate to say they will include "the operating system".

Reviewer 3: I also like this suggestion, as just containerizing the application and it's dependencies is far preferable in my opinion to starting with the state of a whole operating system and adding an application on top. If you're directly dealing with just your application and it's dependencies the state, and changes to it become much more explicit, which can make debugging easier, and avoid problems being introduced in unclear ways.

Using images for the container state is one decision, but if how to build those images is out of scope, then basing the images on operating systems should probably be out of scope as well.

Reviewer 2: I agree - this wasn't meant to imply "we'll use `FROM ubuntu` or equivalent, not `FROM scratch`". I think that decision is out of scope for this RFC - the point is just we'll move away from having a shared operating system provided by the runtime.

A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.

The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.

> GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.

Reviewer: We also have most of GOV.UK containerised suitable for production environments from [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests) - example [Dockerfile](https://github.com/alphagov/publishing-api/blob/1585753a899b037d85ad86e2777c9aac7024356c/Dockerfile)

Reviewer 2: ```suggestion
GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)) and uses production-like containers in the [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests). A containerised runtime will narrow the gap between development, test and production.
```

Reviewer 2: Addressed in b102e3849cd99872c04223312b0388b4f4a9871a

### Benefits of this approach

> The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.

Reviewer 4: Is this referring to upgrading the host OS?

Reviewer 3: Given that the "where" these contained processes are going to run is out of scope for the RFC, we could potentially end up not managing the operating system on which they run, so I'm also unsure what this refers to by operating system version.

I do see a reduction in risk though if you can more tightly manage the applications dependencies, rather than knowing it's probably dependent on _some_ Ubuntu packages or bits of Ubuntu.

Reviewer 2: Yeah, this paragraph is based on the (unchecked) assumption that the container would include an operating system as a dependency (a la `FROM ubuntu` or using the `cflinuxfs3` stack that GOV.UK PaaS provides), which would then be easier to upgrade (than the existing infrastructure which has multiple applications sharing the same OS).

The idea is very much that we wouldn't end up managing the host operating system on which the containers run, leaving that to whichever platform we end up with.

I guess we should reword this along the lines of:

```suggestion
The most immediate benefit of this approach is that it will allow us to offload upgrades of the host operating system to whomever manages the container runtime, and only have to manage updates to the applications' dependencies within the containers. By using containers to manage application dependencies more tightly, GOV.UK can enable a much lower risk upgrade path. 
```

Reviewer 3: Personally, I would see this in terms of separation between the applications and the operating system, and having the opportunity to be more specific about what applications require and use. I'm not sure how related this RFC is to offloading responsibility for managing stuff at the operating system level.

Reviewer 3: > The idea is very much that we wouldn't end up managing the host operating system on which the containers run, leaving that to whichever platform we end up with.

I didn't get this from reading the proposal section. In fact the wording suggests otherwise "
GOV.UK will adopt a container based hosting environment", in that GOV.UK will have responsibility for the hosting environment, rather than just "begin using a container based hosting environment".

Reviewer 2: I've suggested a change in wording in the proposal section to clear that up.

I think we'd like to avoid having responsibility for the hosting environment if at all possible.

Reviewer 3: > I've suggested a change in wording in the proposal section to clear that up.
> 
> I think we'd like to avoid having responsibility for the hosting environment if at all possible.

Cool, I see that the proposal is now to "use a managed" platform.

Returning to the discussion of what "incrementally upgrade the operating system version" means in the context of containers, I like the change you proposed Richard. Hopefully if we start working at the "container" (which I read as process) level, and doing this on a managed platform, then there wouldn't be any operating systems to directly interact with.

Reviewer 2: > there wouldn't be any operating systems to directly interact with.

I guess it's a definitional thing. If we're running ruby apps (and let's definitely assume we are), we're going to need at least need glibc (or musl), openssl etc., which are "Operating system" level things.

We can think of these things as dependencies (rather than bits of an OS) if we like, but we'll still have to think about monitoring them for vulnerabilities and patching them and so on.

Given that, we _might_ decide to put an OS distribution in the container, because then we can piggy back on the distribution's patch schedule. Or we might not, because the fewer things we depend on the fewer things we need to patch. Or we might decide to use the cloud foundry approach, and build "droplets" instead of full images, in which case the operating system really doesn't live in the image, but gets added to the container at runtime. All interesting options with their own pros and cons.

I think that decision is out of scope for this RFC though - suffice to say we'll still have to patch things inside the containers, but that should be easier than doing an upgrade of an OS on a machine with multiple apps running on it.

Reviewer 2: I've committed the suggestion above in 9aa241450f1e0365443e30b9e2eb151098c79586, hopefully that clarifies this point enough.

Marking this conversation as "Resolved" to try and clear up the RFC, but let me know if it needs more clarification :)

> Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.

Reviewer 4: Could add that it would allow to test configuration locally? - as we currently no way to test puppet without running it in an environment.

Reviewer 2: I think we're kind of alluding to this below with:

> Running containers in production allows for a reduction in the difference between development environments and production

It would be good to call out being able to test configuration changes a bit more explicitly, although in practice there may still be a fair amount of configuration that's different between development and production (depending on which platform we go with).

Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.

> Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.

Reviewer 6: Does this cater for everyone working on GOV.UK? In theory you'll find bugs faster and more consistently because environments are similar, but different people use different tools when developing and diagnosing bugs. For example, how easily can an IDE be used to debug a Docker application? We might need to do some more investigation and put more effort into this first.

Reviewer 4: Isn't the emphasis that production bugs are found more quickly as they are more likely to show up and be reproduced in your development environment? Regardless of how you go about debugging them.

Reviewer: I'd suggest that this benefit may be rather dependant on the decisions about the type of platform this is hosted on. For example if we use a build pack based approach for production there's likely to be no reduction in differences between dev and prod.

Reviewer 2: Yeah, this point hides a few assumptions about the type of platform, how containers are built, and where they fit in to the development experience, all of which are out of scope for the RFC.

That said, the statement that "Running container in production _allows_ for a reduction in the difference" is still true, even if we decide not to do much with production-like development environments.

@theseanything mentions testing configuration locally, which I think is a good example of the benefit of being able to play with things in a production-like configuration outside of a full environment. Perhaps we could add that to this point:

```suggestion
Running containers in production allows for a reduction in the difference between development environments and production. For example, it can allow developers to test some configuration changes locally, without having to deploy to an environment. This can lead to bugs being found more quickly, and faster development cycles.
```

Reviewer 2: I've added the above suggestion in c8170ef6c0d727b6638d83849a1263c2d5769cfd.

I'm not sure that fully resolves @karlbaker02 and others comments in this thread though.

I'd say that this RFC isn't proposing a change to the experience of writing code or debugging local apps - people would still do that with govuk-docker or however they personally prefer. That process won't get any better with containers in production.

What _might_ get better is the ability to do `docker-compose up` (or similar) with the versions of the containers that are run in production. As @kevindew says, whether that's really practicle rather depends on some choices we haven't made yet (e.g. currently buildpacks are less suitable for this use case).

I think the current wording ("allows for a reduction") is vague enough to suggest that this is a benefit we may or may not see depending on other choices.

Does that resolve this question?

Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.

> ### Risks of this approach

Reviewer 7: I think the reason we have so much outdated infra (OS, puppet, fabric, etc) is because it works well enough for us to not prioritise updating it.  And now it's become a big problem.

What guarantee do we have that any container stuff we implement won't become similarly outdated?

Author: I don't think we have a guarantee, but if we can offload the need to support the OS and related dependencies to the platform provider then it should give us more time to keep any container stuff up to date.

Reviewer 8: I think that one of the benefits of using containers is that it breaks the big problems down in to smaller ones.

The OS, fabric and puppet problems are all big because they touch so many things, which creates inertia to address the problem, and then deprioritisation.

Compare it with containers:  lets say SomeLib needs updating from v2 to v3 in 20 containerised apps.  We can now break it in to 20 smaller pieces of work with different priorities; making the problem smaller also means we can have higher confidence and lower impact from things like dependabot auto-upgrading libraries and auto feeding that auto-upgrade in to the testing pipeline.

Where I am going with this all is: i think containers will make the work smaller, easier, routine and possibly even dissapear if we can automate it enough.  Meaning prioritisation will become less of a problem and help mitigate this problem appearing again

Reviewer 2: Added these sentences to the RFC in 101523a :

> Future upgrades should be easier to prioritise, because they will be smaller, lower risk pieces of work.

> GOV.UK will still need to continuously upgrade some infrastructure dependencies. There is a risk that this work will not be prioritised, leaving GOV.UK with a different set of out-of-date infrastructure in the future. Moving to managed platforms wherever possible helps mitigate this, but there will always be some maintenance work that needs to be prioritised.

> GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.

Reviewer 7: How long has the AWS migration been happening for?  I'm sure it was being talked about (if not begun) when I started, which was almost two years ago now.  However, the AWS migration feels much simpler to me than switching 100% to containers not running on Trusty for all of GOV.UK.

Reviewer 9: The GOV.UK Infrastructure team moved to Reliability Engineering in October of 2017. So longer than that.

Reviewer 2: > However, the AWS migration feels much simpler to me than switching 100% to containers not running on Trusty for all of GOV.UK.

I'm not sure they're entirely comparable. The AWS migration involved data migrations and tricky bits of networking, which I think would be much less difficult in the move to a container platform.

We've learned a lot of lessons from that migration too, so hopefully future work will got a bit faster.

There's also Verify's history of very difficult VMWare -> AWS migrations, which were followed by a relatively quick move to containers. GOV.UK won't be exactly the same, of course.

So the assumption is that this doable in much less than two years, but that's a risky assumption for sure. It would be fairly bad news if we spent a long time on this work, only to realise that we can't complete it before the end of Trusty ESM.

> GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.

Reviewer: ```suggestion
GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2nd line.
```

> 

Reviewer 2: Responding to @issyl0's comments above (and shamelessly paraphrasing some of the words):

```suggestion

The current migration (lift-and-shift to AWS) has been a complex and challenging piece of work. This has led to several engineers reporting feelings of burnout and exhaustion. If we fail to learn the lessons that GDS' past migration experiences have to offer there is a risk that doing this work could have significant negative consequences for the team, and for individuals' well-being. We should carefully consider how we structure teams, and how we can ensure that the teams have the time to do the work properly, get the training they need to learn new technologies, and have the agency they need to feel that they fully own their platform.

```

Focusing on containerisation will detract from work to improve other parts of the platform.

GOV.UK has a significant amount of infrastructure outside of the applications themselves. In moving to a container based hosting platform, it’s possible that some existing behaviour may be missed, resulting in bugs.
### Risks of not doing this

> GOV.UK reduces its options of how to mitigate Trusty reaching the end of support so the only option left is upgrading Ubuntu. This is potentially wasted work as once the upgrade is done GOV.UK will want to consider if running VMs is best for GOV.UK.

Reviewer 4: Shouldn't this RFC be the consideration that "is running VMs is best for GOV.UK", rather than an upgrade path for Trusty?

Reviewer 2: Could you expand on this point a bit?

I think the RFC does aim to answer "is running VMs best for GOV.UK?", broadly with "No, it would be best for GOV.UK to run containers on a managed platform of some kind".

A big part of the reason for that is the high difficulty of upgrading the OS on our existing VM estate though (in my opinion).

Reviewer 4: Ah yes, sorry I should have wrote "should we containerise GOV.UK?" instead of "is running VMs is best for GOV.UK?".

but essentially I feel this RFC has been framed as solving issue of the expiring LTS of Trusty and us needing an upgrade path, rather than a more wholistic look at why we'd move containers. Maybe the summary paragraph (line 5) needs broadening? 

This line (58) suggests that we'd reconsider containerisation if this RFC was rejected. (Sorry there's a lot of double negative logic going on here)

Reviewer 2: I've suggested an amendment to line 5 to make it a bit broader.

I think this line (58) is trying to say "A risk of not doing this [move to a containerised platform] is that we'll have to update Ubuntu the hard way (i.e. with puppet), which will be a big chunk of effort that brings limited benefit" - is that any clearer? Or perhaps it's less clear?

Reviewer 4: Yup! Makes more sense!

Reviewer 2: I've tried to clarify this a bit in 347e7a64a410ad5abb1d0cdea1c4efca419cf65b - I didn't quite use the wording above, but hopefully it's a bit clearer now.

Let me know if it still doesn't make sense :)

GOV.UK will also have to tackle a number of difficult tooling upgrades (Puppet, in particular) if GOV.UK wants to return to a situation where GOV.UK does not use unsupported dependencies.

> GOV.UK will be harder for the rest of GDS to support, because its infrastructure will be significantly different. Almost all other service teams in GDS are using a container based hosting environment of some kind (mostly [CloudFoundry](https://www.cloudfoundry.org) or AWS ECS).

Reviewer 4: Maybe I've misread this - but isn't doesn't containerisation make GOV.UK more aligned with the rest of GDS, hence easier to support?

Reviewer 10: This paragraph is in the "risks of _not_ doing this" section. You're right: containerisation means more alignment with other parts of GDS, therefore a risk of not doing this is continued/increasing differentiation between us and other areas.

### Follow up work

GOV.UK will commit to doing these things:

> - Investigate the practicalities involved in using different containerised runtimes to host GOV.UK (for example, how difficult will it be to run a hybrid containerised / non-containerised environment in [GOV.UK PaaS](https://www.cloud.service.gov.uk) vs. say [AWS ECS](https://aws.amazon.com/ecs/))

Reviewer 6: Another question is _is PaaS the right place to host containerised apps?_ Whilst it does support Docker, is it the best fit for this?

Reviewer 2: Technically speaking, all the apps that run on GOV.UK PaaS are "containerised" - buildpacks are just a way of building containers, which GOV.UK PaaS then runs.

There's clearly a valid question along the lines of "is GOV.UK PaaS the right place to run GOV.UK's applications?", but we're trying to keep that out of scope for this RFC, and just agree that _some kind of containerised platform_ is where we want to go.

> - Understand how containerising will affect our CI/CD, metrics & alerting, and support models.

Reviewer 3: We don't do CD as far as I know (regardless of whether this means Continuous Delivery or Continuous Deployment). Are you saying there's a commitment to understand how Continuous D(something) would be done?

Reviewer 2: ```suggestion
- Understand how containerising will affect our build and deployment process, observability, alerting, and support models.
```

Reviewer 2: > Are you saying there's a commitment to understand how Continuous D(something) would be done?

I don't think so.

We certainly need to make sure that we can deploy to whatever platform we end up with, and it needs to *at least* not be worse than the current situation.

There's an assumption in this RFC that a container-based platform would make CD easier to implement.

I think the actual decision on whether we want to do CD as part of this work, or whether we want to do it at all, still needs making (and needs to include technical, organisational and human aspects).

Reviewer 2: Turns out we've already got [an RFC for CD accepted](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-070-path-towards-continuous-deployment-cd.md). But this RFC isn't trying to say any more on that subject. I've cleared up the wording not to mention CD in 4d3e92f47b46b679d565891744a7b1361aa43a78.

Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).

Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:

- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.

> - Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.

Reviewer 9: There are ongoing questions about the need for Fabric scripts and how many are regularly in use - @thomasleese knows more, but Platform Health looked at upgrading them to Python 3 from Python 2 is basically a complete rewrite.

Reviewer 2: I know that upgrading verify-fabric was a major blocker for the Verify operating system upgrade, which is why this is called out explicitly. If we can already get rid of it (before the move to containers) that makes this point less compelling. Nevertheless, with the world as it is now, it's a thing that would make an OS upgrade more difficult.

Reviewer 8: If we were to move to a containerised system the need for an elaborate orchestration tool would be mostly removed.

My feeling is that we have been using fabric to prop up our deployment method because it allows us to created complex orchestration procedures.  

Containerisation tech either manages-out or changes deployment strategy which means the complex orchestration rules/conditionals we created in fabric aren't needed.

## Proposal

GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.

GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.

> A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](

Reviewer 9: Did I miss a meeting? This is the first time I've seen the linked document.

Reviewer 10: This is a document I created about 3 years ago, which we updated 6–9 months ago. We could do more in terms of communicating this... Let's say this is part of that!

https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.

The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.

> GOV.UK has already containerised it’s development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.

Reviewer 9: ```suggestion
GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
```

### Benefits of this approach

The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.

Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.

Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.

Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.

> Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.

Reviewer 9: 💯

### Risks of this approach

GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.

> GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.

Reviewer 9: Given this long timeframe, what are we doing and **what will we do to ensure the ongoing health of the _people_ who work on the teams doing this migration**? Is team structure going to change? Will there be an embedded infrastructure team again?

Reviewer 9: This also feels like it will require investment of all devs - not just SREs - to do this properly, with consent, training, etc, and for devs to feel like they have agency in the future of GOV.UK's platform - because that's important because they're the ones supporting it.

Reviewer 6: Building on @issyl0's point, we should also think about skills required during recruitment, as well as the potential need to upskill others within GDS should they switch to working on GOV.UK for whatever reason.

Reviewer 9: Currently we have to upskill them in Puppet, whereas other parts of GDS are already container based (Verify, PaaS), so I'd argue that there's _less_ divergeance here.

Reviewer 2: Yeah, big 👍  to this - the impact on people is a key risk here, and we should call it out. GitHub won't let me make a suggestion in this thread. I'll add one below.

Reviewer 9: Richard's suggestion on wording is in https://github.com/alphagov/govuk-rfcs/pull/117#discussion_r386425034, for those who read this thread and don't want to scroll forever to find it.

Author: It's now part of the RFC 7327c7e4cce50ac8f11fd4de656c106d0934ce84

> Focusing on containerisation will detract from work to improve other parts of the platform.

Reviewer 9: Such as?

Reviewer 2: I think it's just that we'll need some people to be working on containerisation rather than doing other work (like migrating stuff from whitehall to content-publisher, working on the [application architecture goals](https://docs.google.com/document/d/1jEjPOFJ2s1cjQv9vHAbE-KF68LSletpnUVaG5lXlHy4/edit), doing all the other important things we need to do). We have enough people to be doing all of these things, but some things will have to have fewer people working on them if we work on this.

Maybe:

```suggestion
Focusing on containerisation will mean fewer people will be available to work on improving other parts of the platform.
```

Reviewer: @richardTowers you seem to be hinting here that GOV.UK teams will be working on this? Just wondering if that is intentional or if it's not known yet whether this will be GOV.UK teams, RE or a combination of both?

Author: @kevindew There's no plan yet, this RFC is the start of the process. But my guess is that it'll be a combination of both.

Reviewer 9: Relates to https://github.com/alphagov/govuk-rfcs/pull/117#discussion_r386425034 - the follow-up comment from the previous comment (oh no) - about devs having agency in these plans and the work for it.

GOV.UK has a significant amount of infrastructure outside of the applications themselves. In moving to a container based hosting platform, it’s possible that some existing behaviour may be missed, resulting in bugs.
### Risks of not doing this

> GOV.UK reduces it’s options of how to mitigate Trusty reaching the end of support so the only option left is upgrading Ubuntu. This is potentially wasted work as once the upgrade is done GOV.UK will want to consider if running VMs is best for GOV.UK.

Reviewer 9: ```suggestion
GOV.UK reduces its options of how to mitigate Trusty reaching the end of support so the only option left is upgrading Ubuntu. This is potentially wasted work as once the upgrade is done GOV.UK will want to consider if running VMs is best for GOV.UK.
```

> # RFC 117: Use a containerised runtime

Reviewer 3: ```suggestion
# RFC 117: Use a container runtime
```

While I don't have a strong understanding of "container" or "runtime", I think "container runtime" better reflects what this RFC is actually about.

I think most people would say a "containerised app" is an app deployed or running through a container. So a "containerised runtime" is a runtime deployed or running through a container, which I think isn't what this RFC is about.

Reviewer 3: Thanks for looking at this @richardTowers , would it be possible to update other occurrences of "containerised runtime" within the text?

Reviewer 2: Yeah - done in 898f8b35af (I haven't renamed the file though, because I'm concerned that will bamboozle GitHub - we can do that before merging if we remember)

Reviewer 2: Ah, Erin's done the rename. Looks like GitHub is still okay 😸

Author: I've just renamed the file in 952877e and didn't think about breaking GitHub 😟

GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.

A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.

The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.

GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.

> ### Benefits of this approach

Reviewer 3: ```suggestion
### Benefits of this approach

A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.

The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.

GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.

```

While the current first two paragraphs contain proposals, but the other three paragraphs don't contain proposals as far as I can see, but seem to describe perceived benefits, so it would be good to move them to the relevant section making the proposals clearer.

Reviewer 2: Addressed in 5e0871e79290661e1439c486c0e3a439e5f13799
